bundles/org.openhab.binding.solarwatt/src/main/java/org/openhab/binding/solarwatt/internal/handler/EnergyManagerHandler.java

```diff
@@ -0,0 +1,408 @@
+     *
+     * @param solarwattChannel channel description with name and unit
+     */
+    protected void assertChannel(SolarwattChannel solarwattChannel) {
+        ChannelUID channelUID = new ChannelUID(this.getThing().getUID(), solarwattChannel.getChannelName());
+        ChannelTypeUID channelType = this.channelTypeProvider.assertChannelType(solarwattChannel);
+        if (this.getThing().getChannel(channelUID) == null) {
+            ThingBuilder thingBuilder = this.editThing();
+            thingBuilder.withChannel(
+                    SimpleDeviceHandler.getChannelBuilder(solarwattChannel, channelUID, channelType).build());
+
+            this.updateThing(thingBuilder.build());
+        }
+    }
+
+    /**
+     * Finds the guid of the energy manager inside of the known devices.
+     *
+     * @param devices list with known devices
+     * @throws SolarwattConnectionException if there is no energy manager available
+     */
+    private void findEnergyManagerGuid(Map<String, Device> devices) throws SolarwattConnectionException {
+        devices.forEach((guid, device) -> {
+            if (device instanceof EnergyManager) {
+                this.energyManagerGuid = guid;
+            }
+        });
+
+        if (this.energyManagerGuid == null) {
+            throw new SolarwattConnectionException("unable to find energy manager");
+        }
+    }
+
+    /**
+     * Setup the handler and trigger initial load via {@link EnergyManagerHandler::refreshDevices}.
+     *
+     * Web request against energy manager and loading of devices is deferred and will send the ONLINE
+     * event after loading all devices.
+     */
+    @Override
+    public void initialize() {
+        this.logger.debug("{} initialize", this);
+        SolarwattConfiguration localConfig = this.getConfigAs(SolarwattConfiguration.class);
+        this.initRefresh(localConfig);
+        this.initDeviceCache(localConfig);
+    }
+
+    private void initDeviceCache(SolarwattConfiguration localConfig) {
+        if (localConfig.hostname.isEmpty()) {
+            this.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Hostname is not set");
+        } else {
+            this.updateStatus(ThingStatus.ONLINE, ThingStatusDetail.CONFIGURATION_PENDING,
+                    "Waiting to retrieve devices.");
+            this.connector.setConfiguration(localConfig);
+
+            this.devicesCache = new ExpiringCache<>(Duration.of(localConfig.refresh, ChronoUnit.SECONDS),
+                    this::refreshDevices);
+
+            ExpiringCache<Map<String, Device>> localDevicesCache = this.devicesCache;
+            if (localDevicesCache != null) {
+                // trigger initial load
+                this.scheduler.execute(localDevicesCache::getValue);
+            }
+        }
+    }
+
+    /**
+     * Stop the refresh job and remove devices.
+     */
+    @Override
+    public void dispose() {
+        this.logger.debug("{} dispose", this);
```

Comment:
Suggest you remove this, see https://www.openhab.org/docs/developer/guidelines.html#f-logging for why and my comments made already on logging.

Reply:
Excessive logging is removed.